Added build step for generating Windows MSI

james-bebbington opened the pull request:
**Link to tracking Issue:**
https://github.com/open-telemetry/opentelemetry-collector/issues/1125

**Description:**
Added commands & build steps to generate an MSI.

* Running `.\scripts\make.ps1 New-MSI -Config "./path/to/conf.yaml"` will create an MSI packaged with the specified config file. If the -Config argument is not provided, the default config file in `examples/otel-local-config.yaml` will be used.
* An MSI with the default config will be published as a release artifact.

Users are able to override the config when installing the MSI by passing in a different config file path as an argument, i.e.
```cmd
msiexec /i opentelemetry-collector.msi CONFIG=path/to/conf.yaml
```

If the user specifies a custom config file, and that file can't be found, they will get a generic popup error. The MSI logs will report "CustomAction CopyConfig returned actual error code 4" which is a pretty mediocre error message, but at least points to the error.

(once this gets merged, I'll create a similar PR in Contrib)

james-bebbington commented on .circleci/config.yml:
```diff
@@ -73,7 +73,7 @@ workflows:
       - cross-compile:
           requires:
-            - build
+            - setup-and-lint
```
I'm not sure if there was any reason why the `cross-compile` step was waiting for the `build` step previously?

I've made the `windows-msi` step run after the `cross-compile` so that it can just make use of the already generated binary & doesn't need to upgrade Go etc. This step can take a couple of minutes and I didn't want to this have to be a fourth step in the pipeline.

i.e.

![image](https://user-images.githubusercontent.com/568630/85813198-ed7b7780-b7a5-11ea-8349-bbff372154e3.png)
vs
![image](https://user-images.githubusercontent.com/568630/85813238-02f0a180-b7a6-11ea-9c8d-355924026c02.png)

Having said that, if we only run this step after merges, maybe this is a moot point.

james-bebbington replied:
@owais 

I think this PR is ready to be merged, however, should I make the same change here as you made in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/384/files and replace build with cross-compile? (I had changed this dependency to speed up the build since this PR adds another step that's dependeny on cross-compile)

james-bebbington commented on .circleci/config.yml:
```diff
@@ -284,3 +290,23 @@ jobs:
+          name: Install Wix Toolset
+          command: |
+            choco install wixtoolset
+            setx /m PATH "%PATH%;C:\Program Files (x86)\WiX Toolset v3.11\bin"
```
Annoyingly we have to do this because the choco package doesn't take care of it for us: https://chocolatey.org/packages/wixtoolset#comment-3471097677

james-bebbington commented on Makefile:
```diff
@@ -281,3 +281,17 @@ genproto_sub:
 	go run cmd/pdatagen/main.go
+
+# The following commands can only be run on Windows (use WSL2 or Cygwin,
+# or run equivalent commands directly in Cmd/Powershell)
```
Not sure if its worth adding this? It's very useful to have a command that you can run locally on Windows to build an msi, especially to support the first bullet point in https://github.com/open-telemetry/opentelemetry-collector/issues/1125. However, the Wix Toolset only works on Windows, so these commands will not work on Linux/Mac. And make commands are not the best way to define this on Windows either (although if WSL2 goes GA soon then maybe this is okay).

Side note: I experimented with [wixl](https://wiki.gnome.org/msitools) for quite a while but it doesn't support all wix components, such as CopyFile or most CustomAction elements, and wasn't worth the effort in the end.

jchengsfx replied:
@james-bebbington Yes, I suggest moving this to a separate powershell script, something like `scripts/msi/make.ps1`.  Then we can run it as `.\scripts\msi\make.ps1 -version VERSION -config CONFIG`.

jrcamp commented:
> I need a bit of guidance with this PR. Please see the questions below:
> 
> 1. When should this step run? Shall we set this up similarly to the `publish-dev` & `publish-stable` steps? (can't be combined into the same steps as will need to run under a diff executor unfortunately)
> 2. Where should we store the generated MSI? Afaik, there isn't really any standard package manager for storing MSIs. Is storing it as a release artifact enough? That would mean we only run this on `publish-stable`. Should we run a step to create the MSI on merge anyway, so we find out sooner if this somehow gets broken?
> 3. Also see question below re Makefile

Do workspaces persist across executor types? I forget. If so then maybe you can add windows-msi to requires of publish-stable? And in the job store the msi using store_artifacts?

> That would mean we only run this on `publish-stable`. Should we run a step to create the MSI on merge anyway, so we find out sooner if this somehow gets broken?

Not sure what you mean by this. It was run as part of the circleci job in this PR for instance even though it's not part of publish-stable right?

jrcamp commented:
btw would be nice to get into chocolatey repo (@jchengsfx did it for smart agent recently)

james-bebbington commented:
> @james-bebbington Yes, I suggest moving this to a separate powershell script, something like `scripts/msi/make.ps1`. Then we can run it as `.\scripts\msi\make.ps1 -version VERSION -config CONFIG`.

Cool - done. I've set it up fairly similarly to the signalfx script. Also update the `windows-test` step to use this script as well

> Do workspaces persist across executor types? I forget. If so then maybe you can add windows-msi to requires of publish-stable? And in the job store the msi using store_artifacts?

Yes - done (I tested to make sure `persist_to_workspace` was working by creating a temporary build step that I've now deleted)

> > That would mean we only run this on `publish-stable`. Should we run a step to create the MSI on merge anyway, so we find out sooner if this somehow gets broken?
> 
> Not sure what you mean by this. It was run as part of the circleci job in this PR for instance even though it's not part of publish-stable right?

It's running as part of this PR for now, but I don't think we should create an MSI on every build. I was thinking to add the following filter since we only need the MSI when doing a release:
```yaml
branches:
  ignore: /.*/
```

The downside to that is if someone were to break the create MSI build step, we wouldn't find out until the next time someone does a release which would be a little bit annoying.

So instead I've changed the filter to (currently commented out):
```yaml
branches:
  only: master
```

On merge, this will create the MSI and do nothing with it. On release, this will create the MSI and publish it as part of the release tags.

james-bebbington commented:
(this PR is now ready for review/merge)

jchengsfx commented on .circleci/config.yml:
```diff
@@ -101,6 +101,14 @@ workflows:
+            - cross-compile
+          filters:
+            branches:
+              only: master
```
@james-bebbington I think we should remove the branches filter and run for all PRs.  Otherwise, the job won't get executed for PRs (like this one) and we wouldn't have any visibility into it until after it gets merged to master.

jchengsfx commented on .circleci/config.yml:
```diff
@@ -284,3 +294,19 @@ jobs:
+          command: .\scripts\make.ps1 Install-Tools
+      - run:
+          name: Build MSI
+          command: .\scripts\make.ps1 New-MSI
```
@james-bebbington Maybe we can add a simple `msiexec` installation test step in this job after building the msi?  Or better yet, we can add a new job that would test the installation, service start/stop, and uninstallation.

james-bebbington replied:
I included this validation check as part of the same `windows-msi` job. It does feel like it would be nice to do this in another job, but that would require another `attach_to_workspace` step, etc, which is quite slow.

jchengsfx commented:
> btw would be nice to get into chocolatey repo (@jchengsfx did it for smart agent recently)

@jrcamp Building a .nupkg package should be fairly straight-forward.  But getting it into https://chocolatey.org/ would require some coordination with the otel community and governance board since we'll need to register for a chocolatey account (which I'm assuming we don't want an individual to own).  In the short term, we can try building the package in circleci and release to github.  Users can then download it from github and install the downloaded package locally with choco.

jrcamp commented:
@james-bebbington I think it's fine to build msi on every build. Like you say we don't want to find out it's broken after we make a release. It should happen in parallel to other things like running tests so I don't think it should make the build much longer.

james-bebbington commented:
> @james-bebbington Let's rename the `scripts/msi/` directory to `packaging/msi/`, move `make.ps1` to `packaging/msi/`, and update the paths in the files accordingly.

Cool - all done

> btw would be nice to get into chocolatey repo (@jchengsfx did it for smart agent recently)

I'll see how I go for time. Would be nice to add this, but we likely won't need it as we have another [package manager](https://github.com/google/googet) for Windows on GCP.

jchengsfx commented on packaging/msi/make.ps1:
```diff
@@ -0,0 +1,67 @@
+function New-MSI(
+    [string]$Config="./examples/otel-local-config.yaml"
+) {
+    $Version = If($env:VERSION) { $env:VERSION } else { "0.0.1" }
```
We should also set `VERSION` to the `CIRCLE_TAG` env var (if it exists, without the `v`) for release builds.

james-bebbington replied:
Done - since we're referring to a CircleCI specific env variable, I moved this code into the CircleCI config file.

Minor note: MSI [Product Version](https://docs.microsoft.com/en-us/windows/win32/msi/productversion) only supports the regex format you specified (e.g. 0.0.1). The tag filter we use to determine when to publish releases allows additional characters at the end (e.g. v0.0.1-alpha). If we see a tag like that I've opted to just ignore the extra chars.

james-bebbington commented:
I'm very keen to merge this one soon if anyone is able to take another look
